### PR TITLE
8309209: C2 failed "assert(_stack_guard_state == stack_guard_reserved_disabled) failed: inconsistent state"

### PR DESCRIPTION
**Pull request by dholmes-ora**
This appears to be the same kind of issue as reported in [JDK-8146697](https://bugs.openjdk.org/browse/JDK-8146697) way back in Java 9, which was only "fixed" on x86.  The current failure was seen on Aarch64. It seems prudent to apply the same changes to all the other platforms. I've done Aarch64, and took a guess at RISC-V but do not know PPC or S390, so I am looking to others to provide the appropriate equivalent code changes there.

Testing so far is Aarch64 only:
- Tiers 1-3
- 50x the closed stackoverflow test that failed previously
- 25x vmTestbase/nsk/stress/stack/*

As these failures are so rare, passing tests don't really tell us much. This is more an attempt at additional robustness.

Thanks.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8309209](https://bugs.openjdk.org/browse/JDK-8309209): C2 failed "assert(_stack_guard_state == stack_guard_reserved_disabled) failed: inconsistent state" (**Bug** - P3)

### Reviewers
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - **Reviewer**) ⚠️ Review applies to [1d001c3b](https://git.openjdk.org/jdk/pull/14669/files/1d001c3b91864263c9d1a8539077a232fc34535e)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [1d001c3b](https://git.openjdk.org/jdk/pull/14669/files/1d001c3b91864263c9d1a8539077a232fc34535e)

### Contributors
 * Fei Yang `<fyang@openjdk.org>`
 * Martin Doerr `<mdoerr@openjdk.org>`
 * Amit Kumar `<amitkumar@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14669/head:pull/14669` \
`$ git checkout pull/14669`

Update a local copy of the PR: \
`$ git checkout pull/14669` \
`$ git pull https://git.openjdk.org/jdk.git pull/14669/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14669`

View PR using the GUI difftool: \
`$ git pr show -t 14669`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14669.diff">https://git.openjdk.org/jdk/pull/14669.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14669#issuecomment-1608866044)

### CODE REVIEW

**Review by theRealAph — APPROVED** [review]

**Comment by dholmes-ora** [comment]
Thanks for the review @theRealAph !

File: src/hotspot/cpu/riscv/interp_masm_riscv.cpp [hunk 1]
```diff
@@ -764,6 +764,10 @@ void InterpreterMacroAssembler::remove_activation(
     // testing if reserved zone needs to be re-enabled
     Label no_reserved_zone_enabling;
 
+    // check if already enabled - if so no re-enabling needed
+    ld(t0, Address(rthread, JavaThread::stack_guard_state_offset()));
+    beq(t0, StackOverflow::stack_guard_enabled, no_reserved_zone_enabling);
+
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/interp_masm_riscv.cpp:770]: Hi, Thanks for handling RISC-V. But this won't build on RISC-V. We would need something like:

```
diff --git a/src/hotspot/cpu/riscv/interp_masm_riscv.cpp b/src/hotspot/cpu/riscv/interp_masm_riscv.cpp
index edec2e08c83..99906c0fea6 100644
--- a/src/hotspot/cpu/riscv/interp_masm_riscv.cpp
+++ b/src/hotspot/cpu/riscv/interp_masm_riscv.cpp
@@ -764,6 +764,11 @@ void InterpreterMacroAssembler::remove_activation(
     // testing if reserved zone needs to be re-enabled
     Label no_reserved_zone_enabling;

+    // check if already enabled - if so no re-enabling needed
+    ld(t0, Address(xthread, JavaThread::stack_guard_state_offset()));
+    sub(t0, t0, StackOverflow::stack_guard_enabled);
+    beqz(t0, no_reserved_zone_enabling);
+
     ld(t0, Address(xthread, JavaThread::reserved_stack_activation_offset()));
     ble(t1, t0, no_reserved_zone_enabling);
```

**Review by RealFYang — CHANGES_REQUESTED** [review]

**Comment by TheRealMDoerr** [comment]
PPC64 implementation:
```diff
--- a/src/hotspot/cpu/ppc/interp_masm_ppc_64.cpp
+++ b/src/hotspot/cpu/ppc/interp_masm_ppc_64.cpp
@@ -888,6 +888,11 @@ void InterpreterMacroAssembler::remove_activation(TosState state,
     // Test if reserved zone needs to be enabled.
     Label no_reserved_zone_enabling;
 
+    // check if already enabled - if so no re-enabling needed
+    lwz(R0, in_bytes(JavaThread::stack_guard_state_offset()), R16_thread);
+    cmpwi(CCR0, R0, StackOverflow::stack_guard_enabled);
+    beq_predict_taken(CCR0, no_reserved_zone_enabling);
+
     // Compare frame pointers. There is no good stack pointer, as with stack
     // frame compression we can get different SPs when we do calls. A subsequent
     // call could have a smaller SP, so that this compare succeeds for an
```

Is the RISCV version correct? `StackGuardState` is an `enum` and should typically have 4 Bytes.

**Comment by dholmes-ora** [comment]
Thanks @TobiHartmann for the review. Note there is no arm32 version here as for some reason it does not have the reserved stack access support, at least in this area.

@RealFYang - thanks for that I will apply your change.

@TheRealMDoerr - thanks for PPC code. I will look into the issue with using the enum as for x64.

**Comment by dholmes-ora** [comment]
> The other implementation don't look correct to me. StackGuardState is an enum and should typically have 4 Bytes.

@TheRealMDoerr could you elaborate please. We have the following:
- x86: we use `void cmpl(Address dst, int32_t imm32);`
- Aarch64: we use `cmp(Register Rd, unsigned char imm8)` and cast to `u1`
- RISC-V: we use `sub (Register Rd, Register Rn, int64_t decrement, Register temp = t0);` (perhaps should be `subw` for 32-bit? @RealFYang ?)

**Comment by RealFYang** [comment]
@dholmes-ora : Yes, that makes sense to me. And I will also need to change to use 32-bit load (lw) to get the guard state.

```
diff --git a/src/hotspot/cpu/riscv/interp_masm_riscv.cpp b/src/hotspot/cpu/riscv/interp_masm_riscv.cpp
index edec2e08c83..1e981498fcc 100644
--- a/src/hotspot/cpu/riscv/interp_masm_riscv.cpp
+++ b/src/hotspot/cpu/riscv/interp_masm_riscv.cpp
@@ -764,6 +764,11 @@ void InterpreterMacroAssembler::remove_activation(
     // testing if reserved zone needs to be re-enabled
     Label no_reserved_zone_enabling;

+    // check if already enabled - if so no re-enabling needed
+    lw(t0, Address(xthread, JavaThread::stack_guard_state_offset()));
+    subw(t0, t0, StackOverflow::stack_guard_enabled);
+    beqz(t0, no_reserved_zone_enabling);
+
     ld(t0, Address(xthread, JavaThread::reserved_stack_activation_offset()));
     ble(t1, t0, no_reserved_zone_enabling);
```

**Comment by dholmes-ora** [comment]
All modified platforms are building correctly - see GHA results.

File: src/hotspot/cpu/riscv/interp_masm_riscv.cpp [hunk 2]
```diff
@@ -764,6 +764,11 @@ void InterpreterMacroAssembler::remove_activation(
     // testing if reserved zone needs to be re-enabled
     Label no_reserved_zone_enabling;
 
+    // check if already enabled - if so no re-enabling needed
+    ldw(t0, Address(xthread, JavaThread::stack_guard_state_offset()));
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/interp_masm_riscv.cpp:768]: It's `lw` instead of `ldw`. See my previous comment. Thanks.

**Comment by offamitkumar** [comment]
Here is s390x port:
```
diff --git a/src/hotspot/cpu/s390/interp_masm_s390.cpp b/src/hotspot/cpu/s390/interp_masm_s390.cpp
index 79c758f8b11..7555eb05b1f 100644
--- a/src/hotspot/cpu/s390/interp_masm_s390.cpp
+++ b/src/hotspot/cpu/s390/interp_masm_s390.cpp
@@ -952,6 +952,11 @@ void InterpreterMacroAssembler::remove_activation(TosState state,
     // Test if reserved zone needs to be enabled.
     Label no_reserved_zone_enabling;
 
+    // check if already enabled - if so no re-enabling needed
+    guarantee(sizeof(StackOverflow::StackGuardState) == 4, "unexptected size");
+    z_ly(Z_R0, Address(Z_thread, JavaThread::stack_guard_state_offset()));
+    compare32_and_branch(Z_R0, StackOverflow::stack_guard_enabled, bcondEqual, no_reserved_zone_enabling);
+
     // Compare frame pointers. There is no good stack pointer, as with stack
     // frame compression we can get different SPs when we do calls. A subsequent
     // call could have a smaller SP, so that this compare succeeds for an
```


@TheRealMDoerr I have a 2nd patch as well, Would you please confirm which will be better:

```
diff --git a/src/hotspot/cpu/s390/interp_masm_s390.cpp b/src/hotspot/cpu/s390/interp_masm_s390.cpp
index 79c758f8b11..a6774326286 100644
--- a/src/hotspot/cpu/s390/interp_masm_s390.cpp
+++ b/src/hotspot/cpu/s390/interp_masm_s390.cpp
@@ -952,6 +952,12 @@ void InterpreterMacroAssembler::remove_activation(TosState state,
     // Test if reserved zone needs to be enabled.
     Label no_reserved_zone_enabling;
 
+    // check if already enabled - if so no re-enabling needed
+    guarantee(sizeof(StackOverflow::StackGuardState) == 4, "unexptected size");
+    z_cli(Address(Z_thread, JavaThread::stack_guard_state_offset() + in_ByteSize(sizeof(StackOverflow::StackGuardState) - 1)),
+            StackOverflow::stack_guard_enabled);
+    z_bre(no_reserved_zone_enabling);
+
     // Compare frame pointers. There is no good stack pointer, as with stack
     // frame compression we can get different SPs when we do calls. A subsequent
     // call could have a smaller SP, so that this compare succeeds for an
```

File: src/hotspot/cpu/riscv/interp_masm_riscv.cpp (hunk 2 above)
Comment (dholmes-ora) [reply src/hotspot/cpu/riscv/interp_masm_riscv.cpp]: Fixed

**Comment by dholmes-ora** [comment]
Thanks @offamitkumar ! I've applied the first variant for now.

/contributor add @offamitkumar

File: src/hotspot/cpu/aarch64/interp_masm_aarch64.cpp [hunk 3]
```diff
@@ -693,6 +693,11 @@ void InterpreterMacroAssembler::remove_activation(
     // testing if reserved zone needs to be re-enabled
     Label no_reserved_zone_enabling;
 
+    // check if already enabled - if so no re-enabling needed
+    ldr(rscratch1, Address(rthread, JavaThread::stack_guard_state_offset()));
+    cmp(rscratch1, (u1)StackOverflow::stack_guard_enabled);
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/aarch64/interp_masm_aarch64.cpp:698]: Not ldrw + cmpw?

**Comment by dcubed-ojdk** [comment]
S390 typo: s/unexptected/unexpected/

File: src/hotspot/cpu/aarch64/interp_masm_aarch64.cpp (hunk 3 above)
Comment (dholmes-ora) [reply src/hotspot/cpu/aarch64/interp_masm_aarch64.cpp:698]: I've no idea, I used the code below as the pattern here. @theRealAph reviewed this.

**Comment by dholmes-ora** [comment]
> S390 typo: s/unexptected/unexpected/

Thanks @dcubed-ojdk , I spotted that before comitting.

File: src/hotspot/cpu/aarch64/interp_masm_aarch64.cpp (hunk 3 above)
Comment (dean-long) [reply src/hotspot/cpu/aarch64/interp_masm_aarch64.cpp:698]: ldrw + cmpw does seem more correct, plus an asset that sizeof _stack_guard_state == 4.  ldr+cmp is only going to work for little-endian, and only as long as the alignment padding between _stack_guard_state and _stack_overflow_limit is all zeroes.

Comment (dholmes-ora) [reply src/hotspot/cpu/aarch64/interp_masm_aarch64.cpp]: Changed

**Comment by bulasevich** [comment]
> Note there is no arm32 version here as for some reason it does not have the reserved stack access support, at least in this area.

Yes, 'JEP 270: Reserved Stack Areas for Critical Sections' is not implemented for arm32, so the current change is not applicable to the platform.

**Comment by dholmes-ora** [comment]
If there are no further comments I will integrate on Monday.

**Comment by TheRealMDoerr** [comment]
Thanks for adding the assertions. I think they would be good to have for all platforms.

**Comment by dholmes-ora** [comment]
> Thanks for adding the assertions. I think they would be good to have for all platforms. I have tested it on PPC64, too.

Thanks @TheRealMDoerr . I have made the comment and assertion the same on all platforms.

**Comment by dholmes-ora** [comment]
Re-testing has passed okay.

/integrate

**Pull request closed without merging**